stdlib/public/core/Random.swift

```diff
@@ -104,19 +104,132 @@ extension RandomNumberGenerator {
+    // required by a factor of two for bignums, while matching or reducing
+    // the expected number of bits required for smaller numbers. This is the
+    // algorithm implemented below (the formally-uniform method is not
+    // much more complex to implement and is only a little bit slower, but
+    // there's no reason to do so).
+    //
+    // More intriguing still, this algorithm can be made unconditional by
+    // removing the early out, so that every value computed requires word
+    // size + 64 bits from the stream, which breaks the loop-carried
+    // dependency for fast generators, unlocking vectorization and
+    // parallelization where it was previously impossible. This is an
+    // especially powerful advantage when paired with bitstream generators
+    // that allow skip-ahead such as newer counter-based generators used
+    // in simulations and ML.
+    //
+    // Note that it is _possible_ to employ Lemire's tighter early-out
+    // check that involves a division with this algorithm as well; this
+    // is beneficial in some cases when upperBound is a constant and the
+    // generator is slow, but we do not think it necessary with the new
+    // algorithm and other planned improvements.
+    //
+    // [1] We can actually achieve log₂(upperBound) + ε for any ε > 0 by
+    // generating multiple random samples at once, but that is only of
+    // theoretical interest--it is still interesting, however, since I
+    // don't think anyone has described how to attain it previously.
+    if T.bitWidth < 64 {
+      // Ensure T is at least 64 bits wide, as this simplifies things
+      // somewhat later on, and doesn't cost us anything with the existing
+      // RNG protocol (we might investigate buffered RNGs that let you
+      // draw less than 64 bits at a time at some future point).
+      return T(truncatingIfNeeded: next(upperBound: UInt64(upperBound)))
     }
-    return m.high
+    let (result, fraction) = upperBound.multipliedFullWidth(by: next())
+    // Optional early out: this is a performance win when the generator is
+    // very slow (as with an unbuffered SystemRandomNumberGenerator), but
+    // the unpredictable branch and variable consumption from the bitstream
+    // is a performance hazard for fast generators. We'll keep it for now,
+    // but future changes to allow buffering slow generators should lead
+    // to its removal.
+    if T(fraction) <= 0 &- upperBound { return result }
+    // Compute the product with 64 additional bits of randomness, add that
+    // to the fraction, and adjust result upwards if that sum carries out.
+    let (_, carry) = T(fraction).addingReportingOverflow(
+      multiplyHigh(upperBound, next())
+    )
+    return result + (carry ? 1 : 0)
+  }
+  
+  /// Computes a*b >> 64.
+  ///
+  /// Requires T.bitWidth >= 64.
+  @_transparent @usableFromInline
+  internal func multiplyHigh<T:FixedWidthInteger & UnsignedInteger>(
```

Comment:
Could delete the ABI footprint of this by making it `@_alwaysEmitIntoClient`. (Would also underscore the name.)